### PR TITLE
Crypto refactor (0 / n)

### PR DESCRIPTION
**Pull request by imeckler**
I'm trying to clean up some of the crypto code and comb through to find any remaining bugs. This PR has refactors and bugfixes for

- Pedersen parameter generation
- field extension arithmetic
- curve arithmetic
- pairings

There was one bug in how Pedersen parameters are generated which was benign from a security standpoint.

I separated it by commit.

### CODE REVIEW

File: src/lib/coda_base/gen/gen.ml [hunk 1]
```diff
@@ -14,7 +14,7 @@ module Group = Curve_choice.Tick_backend.Inner_curve
 
 let bigint_of_bits bits =
   List.foldi bits ~init:Bigint.zero ~f:(fun i acc b ->
-      if b then Bigint.(acc + (of_int 2 lsl i)) else acc )
+      if b then Bigint.(acc + (of_int 1 lsl i)) else acc )
```
Comment (cmr) [inline src/lib/coda_base/gen/gen.ml]: 🤦‍♂

File: src/lib/snarky_field_extensions/field_extensions.ml [hunk 2]
```diff
@@ -73,18 +72,53 @@ module Make (F : Intf.Basic) = struct
 
   let one = constant Unchecked.one
 
+  let constant_optimize to_constant unchecked checked x =
```
Comment (mrmr1993) [inline src/lib/snarky_field_extensions/field_extensions.ml]: Nice!

File: src/lib/snarky_field_extensions/field_extensions.ml [hunk 3]
```diff
@@ -73,18 +72,53 @@ module Make (F : Intf.Basic) = struct
...
+      (mk_checked typ typ unchecked check)
+
+  (* 
+  `div_unsafe x y = z` iff `y * z = x`
+
+   So,
+   - if `y != 0`, then we can divide by `y` and get `div_unsafe x y = z = x / y`.
+   - if `y = 0` and `x != 0`, then the equation says `0 * z = x` which means `0 * z != 0`
+     and so we will have an unsatisfiable constraint system.
+   - if `y = 0` and `x = 0`, then `y` can be anything since the equation just says `0 * z = 0`
```
Comment (mrmr1993) [inline src/lib/snarky_field_extensions/field_extensions.ml]: Nit: `y can be anything` -> `z can be anything`

File: src/lib/snarky_field_extensions/field_extensions.ml [hunk 4]
```diff
@@ -185,6 +190,8 @@ struct
 
   let scale' t x = A.map t ~f:(fun a -> F.scale x a)
 
+  let _ = scale' (* potato *)
```
Comment (mrmr1993) [inline src/lib/snarky_field_extensions/field_extensions.ml]: It's not clear to me what this is doing, but I like it :rofl:

File: src/lib/snarky_field_extensions/field_extensions.mli [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+
+  val special_mul : t -> t -> (t, 'a) Impl.Checked.t
+
+  val special_div : t -> t -> (t, 'a) Impl.Checked.t
+
+  val cyclotomic_square :
+    Fq2.t * Fq2.t -> (Fq2.t * Fq2.t, 'a) Fq2.Impl.Checked.t
+
+  val frobenius :
+       ('a * Impl.field Snarky.Cvar.t)
```
Comment (mrmr1993) [inline src/lib/snarky_field_extensions/field_extensions.mli]: These might be clearer as `Snarky.Field.Var.t`s

**Pull request closed without merging**